google/cloud/storage/internal/grpc_client.cc

```diff
@@ -48,60 +59,54 @@ bool DirectPathEnabled() {
                         [](absl::string_view v) { return v == "storage"; });
 }
 
-std::string GrpcEndpoint() {
+Options FillWithDefaultsGrpc(Options options) {
+  options = FillWithDefaults(std::make_shared<oauth2::AnonymousCredentials>(),
+                             std::move(options));
   auto env = google::cloud::internal::GetEnv("CLOUD_STORAGE_GRPC_ENDPOINT");
   if (env.has_value()) {
-    return env.value();
+    options.set<GrpcCredentialOption>(grpc::InsecureChannelCredentials());
+    options.set<EndpointOption>(*env);
   }
-  return "storage.googleapis.com";
-}
-
-std::shared_ptr<grpc::ChannelCredentials> GrpcCredentials(
-    ClientOptions const& options) {
-  auto env = google::cloud::internal::GetEnv("CLOUD_STORAGE_GRPC_ENDPOINT");
-  if (env.has_value()) {
-    return grpc::InsecureChannelCredentials();
+  if (!options.has<GrpcCredentialOption>()) {
+    // Try to avoid calling grpc::GoogleDefaultCredentials() because it tries to
+    // open files and generates warnings in the CI system as they do not exist.
+    options.set<GrpcCredentialOption>(grpc::GoogleDefaultCredentials());
   }
-  if (dynamic_cast<oauth2::AnonymousCredentials*>(
-          options.credentials().get()) != nullptr) {
-    return grpc::InsecureChannelCredentials();
+  if (!options.has<EndpointOption>()) {
+    options.set<EndpointOption>("storage.googleapis.com");
   }
-  return grpc::GoogleDefaultCredentials();
-}
 
-std::shared_ptr<grpc::ChannelInterface> CreateGrpcChannel(
-    ClientOptions const& options) {
-  return CreateGrpcChannel(options, 0);
+  return options;
 }
 
 std::shared_ptr<grpc::ChannelInterface> CreateGrpcChannel(
-    ClientOptions const& options, int channel_id) {
+    Options const& options, int channel_id) {
   grpc::ChannelArguments args;
   args.SetInt("grpc.channel_id", channel_id);
   if (DirectPathEnabled()) {
-    args.SetServiceConfigJSON(R"json({
-      "loadBalancingConfig": [{
-        "grpclb": {
-          "childPolicy": [{
-            "pick_first": {}
-          }]
-        }
-      }]
-    })json");
-  }
-  return grpc::CreateCustomChannel(GrpcEndpoint(), GrpcCredentials(options),
+    args.SetServiceConfigJSON(kDirectPathConfig);
+  }
+  return grpc::CreateCustomChannel(options.get<EndpointOption>(),
+                                   options.get<GrpcCredentialOption>(),
                                    std::move(args));
 }
 
-GrpcClient::GrpcClient(ClientOptions options)
-    : options_(std::move(options)),
-      stub_(
-          google::storage::v1::Storage::NewStub(CreateGrpcChannel(options_))) {}
+std::shared_ptr<GrpcClient> GrpcClient::Create(Options options) {
+  return Create(std::move(options), /*channel_id=*/0);
+}
+
+std::shared_ptr<GrpcClient> GrpcClient::Create(Options options,
+                                               int channel_id) {
+  // Cannot use std::make_shared<> as the constructor is private.
+  return std::shared_ptr<GrpcClient>(
```

Comment:
Are you planning to call `internal::CheckExpectedOptions()` from somewhere? That should probably be in the first user-facing function that accepts user-provided `options`. I'd imagine that's also the function where you'd call `FillWithDefaults`, which seems like maybe here?

Fine to do in a separate PR.

Reply:
> Are you planning to call `internal::CheckExpectedOptions()` from somewhere?

Yes.

> Fine to do in a separate PR.

That was my plan, I do not know *exactly* what that first entry point will be, there are a bit too many ways to create `storage::Client` things right now.